storage/range_tree.go

```diff
@@ -83,151 +90,178 @@ func getRangeTree(db *client.KV) (*rangeTree, error) {
 	if !ok {
 		return nil, util.Errorf("Could not find the range tree:%s", engine.KeyRangeTreeRoot)
 	}
-	return (*rangeTree)(tree), nil
+
+	return &treeContext{
+		db:          db,
+		tree:        tree,
+		treeUpdated: false,
+		nodes:       map[string]cachedNode{},
+	}, nil
+}
+
+// setRoot sets the tree root key in the cache. It also marks the root for
+// writing during a flush.
+func (t *treeContext) setRootKey(key *proto.Key) {
+	t.tree.RootKey = *key
+	t.treeUpdated = true
 }
 
-// getRangeTreeNode return the RangeTree node for the given key.  If the key is
-// nil, an empty RangeTreeNode is returned.
-func getRangeTreeNode(db *client.KV, key *proto.Key) (*rangeTreeNode, error) {
+// setNode sets the node in the cache so all subsequent reads will read this
+// upated value.  It also marks the node for writing during a flush.
+func (t *treeContext) setNode(node *proto.RangeTreeNode) {
+	t.nodes[string(node.Key)] = cachedNode{
+		node:    node,
+		updated: true,
+	}
+}
+
+// getNode return the RangeTreeNode for the given key.  If the key is nil, nil
+// is returned.
+func (t *treeContext) getNode(key *proto.Key) (*proto.RangeTreeNode, error) {
 	if key == nil {
 		return nil, nil
 	}
+
+	// First check to see if we have the node cached.
+	keyString := string(*key)
+	cached, ok := t.nodes[keyString]
+	if ok {
+		return cached.node, nil
+	}
+
+	// We don't have it cached so fetch it and add it to the cache.
 	node := &proto.RangeTreeNode{}
-	ok, _, err := db.GetProto(engine.RangeTreeNodeKey(*key), node)
+	ok, _, err := t.db.GetProto(engine.RangeTreeNodeKey(*key), node)
 	if err != nil {
 		return nil, err
 	}
 	if !ok {
-		return nil, util.Errorf("Could not find the range tree node:%s",
-			engine.RangeTreeNodeKey(*key))
+		return nil, util.Errorf("Could not find the range tree node:%s", engine.RangeTreeNodeKey(*key))
 	}
-	return (*rangeTreeNode)(node), nil
+	t.nodes[keyString] = cachedNode{
+		node:    node,
+		updated: false,
+	}
+
+	return node, nil
 }
 
 // getRootNode returns the RangeTree node for the root of the RangeTree.
-func (t *rangeTree) getRootNode(db *client.KV) (*rangeTreeNode, error) {
-	return getRangeTreeNode(db, &t.RootKey)
+func (t *treeContext) getRootNode() (*proto.RangeTreeNode, error) {
```

Comment:
It really doesn't feel like this function is useful. Why not just remove it and where it's currently called, invoke t.getNode(&t.tree.RootKey)?